ENH: Use faster solvers for firls

larsoner opened the pull request:
`Q` is symmetric, so we should use that to speed up solving. Also, we can try using `linalg.solve` first, and fall back to a `pinvh` if necessary.

I'm not 100% sure that it's okay to use `solve` in cases where it warns about rank deficiency. It seems to produce okay results in my tests, but if someone thinks it's better to catch the emitted warning and use `pinvh` in that case, too, I can change it to do so.

Closes #9835.

nmayorov commented:
Not very related to the PR (sorry), but does anyone has a suggestion on how to fix the test failure on pypy3?

rgommers commented:
The pypy issue is gh-9837, should be fixed already (to remove the failure here this PR needs rebasing, but you can simply ignore it)

```
______ build/testenv/site-packages/scipy/io/matlab/tests/test_streams.py _______
[gw0] linux -- Python 3.6.1 /home/circleci/project/pypy3.6-latest/bin/pypy3
Worker 'gw0' crashed while running 'build/testenv/site-packages/scipy/io/matlab/tests/test_streams.py::TestZlibInputStream::test_seek'
```

rgommers commented:
> I'm not 100% sure that it's okay to use `solve` in cases where it warns about rank deficiency. It seems to produce okay results in my tests, but if someone thinks it's better to catch the emitted warning and use `pinvh` in that case, too, I can change it to do so.

not quite sure either. @ilayn any thoughts on this one?

ilayn commented:
If singularity(linalgerror)/ill-conditioned(linalgwarning) condition is rare enough, truncating SVD is numerically better at rounding and normalization handling even though they are equal to eigenvalue decomposition theoretically as noted in the comments. If it is common enough you can get the SVD anyways and do it in one shot regardless of the ill-conditionedness.

If you want to keep it as a two branch solution in the second branch I would suggest using `linalg.lstsq` as it avoids the formation of the pseudoinverse. Here is some super quick experiment

```python
n = 10
k = 100
bools = np.empty(k, dtype=bool)
from scipy.linalg import qr, norm, lstsq, pinvh
from numpy.random import rand

for ind in range(k):
    # Generate a random ill-conditioned sym. matrix by garbling it with an ortho. matrix
    zzz = rand(n)
    # Make it numerically ill-conditioned
    zzz -= zzz[3] - 1e-8
    q, _ = qr(np.random.rand(n, n))
    A = q.T @ (zzz * q)
    b = rand(n)
    bools[ind] = norm(A @ (lstsq(A, b)[0]) - b) <= norm(A @ (pinvh(A) @ b) - b)

# Print how many times pinvh outperformed out of k tries
print(bools[~bools].size)
```

larsoner commented:
>  If it is common enough you can get the SVD anyways and do it in one shot regardless of the ill-conditionedness.

It shouldn't be too common, and computing the SVD is much (as in at least an order of magnitude) slower than using `linalg.solve` (for the problems `solve` works for).

> However to enter the second branch you need to catch the LinAlgWarning instead since hitting LinAlgError is pretty rare.

In testing these particular use cases it seemed like the error was more common, but regardless I agree it would be better to catch both.

larsoner commented:
@ilayn you were right, `lstsq` was faster even though it does not AFAIK make use of the symmetry of the problem. I added some code to catch the ill-conditioned warning, too. So hopefully I've addressed all comments at this point.

ilayn commented:
Yes +1 for merge from me.

larsoner commented:
Benchmark added. On `master`:
```
[100.00%] ··· Running signal.FIRLS.time_firls                                                                                                                                     ok
[100.00%] ····
               ====== ============ ==============
               --                edges
               ------ ---------------------------
                 n     (0.1, 0.9)   (0.01, 0.99)
               ====== ============ ==============
                 21     397.52μs      384.81μs
                101     834.15μs      868.89μs
                1001    93.97ms       102.93ms
                2001    958.26ms      951.05ms
               ====== ============ ==============
```
On this PR (note big difference in last two rows in particular a factor of ~5):
```
              ====== ============ ==============
               --                edges
               ------ ---------------------------
                 n     (0.1, 0.9)   (0.01, 0.99)
               ====== ============ ==============
                 21     330.67μs      431.28μs
                101     580.09μs      575.02μs
                1001    20.66ms       20.72ms
                2001    194.04ms      206.35ms
               ====== ============ ==============
```
I ultimately removed the 2001 case because it adds some seconds to the benchmark and doesn't add too much to the story.

rgommers commented:
Looks good to go, merged. Thanks @larsoner and reviewers
